File: src/exception/FException.ts
export class FException extends Error {
	public static wrapIfNeeded(likeError: any): FException {
		if (likeError instanceof FException) {
			return likeError;
		} else if (likeError instanceof Error) {
			return new FExceptionNativeErrorWrapper(likeError);
		} else {
			return new FExceptionNativeObjectWrapper(likeError);
		}
	}

	public readonly innerException: FException | null;

	public constructor();
	public constructor(message: string);
	public constructor(message: string, innerEx?: FException);
	public constructor(innerEx: FException);

	public constructor(first?: any, second?: any) {
		if (first === undefined) {
			super();
			this.innerException = null;
		} else {
			if (typeof first === "string") {
				super(first);
				if (second !== undefined && second !== null) {
					this.innerException = FException.wrapIfNeeded(second);
				} else {
					this.innerException = null;
				}
			} else if (first instanceof FException) {
				super();
				this.innerException = first;
			} else {
				throw new Error("Wrong arguments. First agrument should be a string (if present).");
			}
		}
	}

	public get name(): string {
		return this.constructor.name;
	}

	public toString(): string {
		if (this.innerException !== null) {
			return [super.toString(), this.innerException.toString()].join("\n");
		}
		return super.toString();
	}
}

export class FExceptionNativeObjectWrapper extends FException {
	public readonly nativeObject: any;

	public constructor(nativeObject: any) {
		super(`${nativeObject}`);
		this.nativeObject = nativeObject;
	}
}

export class FExceptionNativeErrorWrapper extends FException {
	public readonly nativeError: Error;

	public constructor(nativeError: Error) {
		super(nativeError.message);
		this.nativeError = nativeError;
	}

	public get name(): string {
		return this.nativeError.name;
	}
}

File: lib/exception/f_exception.dart
// Copyright 2023, the Freemework.ORG project authors. Please see the AUTHORS
// file for details. All rights reserved. Use of this source code is governed
// by a BSD-style license that can be found in the LICENSE file.

class FException extends Error {
  static FException wrapIfNeeded(Object likeError, [StackTrace? stackTrace]) {
    if (likeError is FException) {
      return likeError;
    } else if (likeError is Error) {
      return FExceptionNativeErrorWrapper(likeError);
    } else if (likeError is Exception) {
      return FExceptionNativeExceptionWrapper(likeError);
    } else {
      return FExceptionNativeObjectWrapper(likeError);
    }
  }

  final FException? innerException;

  FException([String? message, this.innerException]) : _message = message;

  String get message {
    return this._message ?? "Exception of type '$runtimeType' was thrown.";
  }

  @override
  String toString() {
    final StringBuffer messageBuffer = StringBuffer();
    final StringBuffer stackTraceBuffer = StringBuffer();

    messageBuffer.write(this.runtimeType);
    messageBuffer.write(": ");
    messageBuffer.write(message);

    FException? innerException = this.innerException;
    while (innerException != null) {
      messageBuffer.write(" ---> ");

      messageBuffer.write(innerException.runtimeType);
      messageBuffer.write(": ");
      messageBuffer.write(innerException.message);

      final StackTrace? innerStackTrace = innerException.stackTrace;
      if (innerStackTrace != null) {
        stackTraceBuffer.write(innerStackTrace.toString());
      } else {
        stackTraceBuffer.writeln("No available stack trace");
      }
      stackTraceBuffer.writeln("--- End of inner exception stack trace ---");

      innerException = innerException.innerException;
    }

    final StackTrace? stackTrace = this.stackTrace;
    if (stackTrace != null) {
      stackTraceBuffer.write(stackTrace.toString());
    } else {
      stackTraceBuffer.write("No available stack trace");
    }

    messageBuffer.writeln();
    messageBuffer.write(stackTraceBuffer.toString());

    return messageBuffer.toString();
  }

  final String? _message;
}

class FExceptionNativeObjectWrapper extends FException {
  final dynamic nativeObject;

  FExceptionNativeObjectWrapper(this.nativeObject)
      : super(nativeObject.toString());
}

class FExceptionNativeErrorWrapper extends FException {
  final Error nativeError;

  FExceptionNativeErrorWrapper(this.nativeError)
      : super(nativeError.toString());

  @override
  String get message => nativeError.toString();

  @override
  StackTrace? get stackTrace => this.nativeError.stackTrace;
}

class FExceptionNativeExceptionWrapper extends FException {
  final Exception nativeException;

  FExceptionNativeExceptionWrapper(this.nativeException)
      : super(nativeException.toString());
}
